service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogHandler.java

```diff
@@ -762,7 +777,190 @@ public LoadTableResponse updateTable(
+                    // lineage
+                    // lets not complicate things rn
+                    throw new ValidationFailureException(e);
+                  }
+
+                  Long parentSnapshotId = addSnapshot.snapshotId();
+                  // so we will first check all the snapshots on the top of
+                  // base on which the snapshot we want to commit is of type REPLACE ops.
+                  Long parentToRollbackTo = base.currentSnapshot().snapshotId();
+                  List<MetadataUpdate> updateToRemoveSnapshot = new ArrayList<>();
+                  while (!Objects.equals(parentToRollbackTo, parentSnapshotId)) {
+                    Snapshot snap = ops.current().snapshot(parentToRollbackTo);
+                    if (!DataOperations.REPLACE.equals(snap.operation())) {
+                      break;
+                    }
+                    updateToRemoveSnapshot.add(
+                        new MetadataUpdate.RemoveSnapshot(snap.snapshotId()));
+                    parentToRollbackTo = snap.parentId();
+                  }
+
+                  MetadataUpdate.SetSnapshotRef ref = null;
+                  found = 0;
+                  // find the SetRefName snapshot update
+                  for (MetadataUpdate update : request.updates()) {
+                    if (update instanceof MetadataUpdate.SetSnapshotRef) {
+                      ++found;
+                      ref = (MetadataUpdate.SetSnapshotRef) update;
+                    }
+                  }
+
+                  if (found != 1 || !Objects.equals(parentToRollbackTo, parentSnapshotId)) {
+                    // nothing can be done as this implies there was a non replace
+                    // snapshot in between or there is more than setRef ops, we don't know where
+                    // to go.
+                    throw new ValidationFailureException(e);
+                  }
+
+                  // first we should also set back the ref we wanted to set, back to the base
+                  // on which the current update is based on.
+                  metadataBuilder.setBranchSnapshot(parentSnapshotId, ref.name());
+
+                  // apply the remove snapshots update in the current metadata.
+                  // NOTE: we need to setRef to parent first and then apply remove as the remove
+                  // will drop. The tags / branch which don't have reference.
+                  // NOTE: we can skip removing the now orphan base. Its not a hard requirement.
+                  // just something good to do, and not leave for Remove Orphans.
+                  updateToRemoveSnapshot.forEach((update -> update.applyTo(metadataBuilder)));
+                  // Ref rolled back update correctly to snapshot to be committed parent now.
+                  newBase = metadataBuilder.build();
+                  // move the lastSequenceNumber back, to apply snapshot properly.
```

Comment:
Why is it important that we move the sequence number back? They are required to be monotonic but not contiguous I believe?

Reply:
Posting offline discussion : seq-number is generatred at client end and sticks with snapshot, since we are rolling back the current base lastSeq number would have gone up, so it should be fine to point this back to the seq number of the snapshot we wanna apply.